### ISSUE TITLE
WebUI: GitHub OAuth app should only require read:user scope

### ISSUE THREAD

**Issue by christopher-dG** [JuliaRegistries/Registrator.jl#290]
Right now it has `public_repo` which is read/write access: https://github.com/JuliaRegistries/Registrator.jl/blob/4777e74a36af39a98e4c413665c2d5763a922763/src/webui/providers.jl#L28
We really only need `read:user`. But first we need to make sure that we're not using the user tokens for stuff that will no longer be allowed with less scopes (I think that the user tokens are used for reading repo data currently).

**Comment by aviks** [comment]
It needs to read the committer list to figure out if the person making the registration has commit rights to the repository. This needs write access to the repo

If we remove that check, you'll have to live with the possibility that any github user can register or tag any package.

Which risk do people want to live with?

**Comment by DilumAluthge** [comment]
> It needs to read the committer list to figure out if the person making the registration has commit rights to the repository. This needs write access to the repo
> 
> If we remove that check, you'll have to live with the possibility that any github user can register or tag any package.
> 
> Which risk do people want to live with?

I don't want any GitHub user to be able to register versions of my package. I'd rather keep the current Registrator.jl permissions.

**Comment by christopher-dG** [comment]
Man, GH's permissions are laughable... :(

**Comment by christopher-dG** [comment]
There are ways that we could get read permissions by default and only prompt for a writable token when needed, but I think it's probably more work than it's worth given how few people use the web interface (especially since this particular issue is GitHub-specific, where users already have a read-only alternative). Also GH OAuth tokens don't expire, so a user that authorized write permissions would need to manually revoke the token to stay "safe" between registrations. Seems like a lot of extra work for little benefit. 🙁

**Comment by StefanKarpinski** [comment]
> It needs to read the committer list to figure out if the person making the registration has commit rights to the repository. This needs write access to the repo

Correct me if I'm wrong, but isn't this only true if membership in the org is not public? If a person is a _public_ member of an org, shouldn't we know that without elevated permissions? If that's the case, then the flow here could be giving them direct link to the page where they can make their membership public and indicate that they must do that in order to be allowed to tag releases.

**Comment by DilumAluthge** [comment]
> > It needs to read the committer list to figure out if the person making the registration has commit rights to the repository. This needs write access to the repo
> 
> Correct me if I'm wrong, but isn't this only true if membership in the org is not public? If a person is a _public_ member of an org, shouldn't we know that without elevated permissions? If that's the case, then the flow here could be giving them direct link to the page where they can make their membership public and indicate that they must do that in order to be allowed to tag releases.

How does this work for repositories that are owned by an individual, not an organization?

**Comment by StefanKarpinski** [comment]
Well, obviously we can let that person tag new releases of their own repo. Is there a public list of who has write access to a personal repo in addition to the owner? But honestly, encouraging people to transfer packages to orgs is a good thing as well. The happy path here is: packages that many people make releases of should belong in orgs and their membership in that org should be public.

**Comment by giordano** [comment]
A member of an organisation may not have write access to all repos of the organisation.

**Comment by DilumAluthge** [comment]
> A member of an organisation may not have write access to all repos of the organisation.

Correct. However, the current behavior of Registrator is: if a user is a member of an organization, and that membership is public, then that user is allowed to register any repo in that organization.

**Comment by giordano** [comment]
Oh, so it doesn't check anyway the actual permissions in the specific repo.

**Comment by DilumAluthge** [comment]
> Oh, so it doesn't check anyway the actual permissions in the specific repo.

It does both. It checks the public members of the org. But also it checks who has push access to the repo. If you are in either group, then it lets you register.

**Comment by christopher-dG** [comment]
There's also outside collaborators to think about.

**Comment by DilumAluthge** [comment]
> There's also outside collaborators to think about.

Yeah. Those should get picked up in the "also it checks who has push access to the repo".

**Comment by christopher-dG** [comment]
Yeah but you can't do that check without push access of your own...

**Comment by DilumAluthge** [comment]
> Yeah but you can't do that check without push access of your own...

Yeah exactly.

**Comment by DilumAluthge** [comment]
Personally I would vote to just leave things the way they are.

**Comment by DilumAluthge** [comment]
Maybe one day GitHub will give us an API for getting a list of users with push access without needing push access ourselves.

**Comment by StefanKarpinski** [comment]
In what situation does an outside collaborator need to be able to register new versions?

**Comment by christopher-dG** [comment]
I'm an outside collaborator on PkgTemplates, I also have repos owned by me with collaborators. I don't think it's very uncommon.

**Comment by StefanKarpinski** [comment]
And I suppose you can't be added to the @invenia org because you're not working there anymore? In that case, it seems like there are a few options:

1. Transfer PkgTemplates into a public org instead of a private one
2. You don't register new PkgTemplates versions, someone at @invenia has to do it
3. Have the _option_ of granting write access, but don't require it

While 3 would be fine, I feel like if no one at a private org is active enough in maintaining a project that they can register new versions, then that project really should be transferred to an org where someone who is a full member can maintain it.

**Comment by StefanKarpinski** [comment]
I realize this is an opinionated value judgement, but I think that registered open source _should_ be maintained by people who are public members of orgs. That's why I'm ok with requiring that. Yes, that's inconvenient if you have a registered package in your personal repo and want help maintaining. However, I think that inconvenience is good because as soon as there's multiple people maintaining a package to the level where others are tagging versions, then the repo should be moved to a public org where all of the maintainers who tag new versions are public members. Same argument with repos like PkgTemplates that are in a private org: if it's really a pain point that someone who's not in the org can't tag new versions, then it's time to move the package.

**Comment by christopher-dG** [comment]
> Have the option of granting write access, but don't require it

If you mean what I think you mean, then this is what I meant to when I mentioned that it was technically possible to only prompt for a writable token when necessary, but like I said it would be kind of complicated not much benefit.

I'm not super invested in this, as I also mentioned anyone taking issue with the permissions is free to use the App since this issue is GitHub-specific. I'm fine with either outcome.
